### ISSUE TITLE
Log Study enrollment/unenrollment using Telemetry Events

### ISSUE THREAD

**Issue by mythmon** [mozilla/normandy#1139]
It is possible to derive data about when a client enrolls and unenrolls from studies by watching the experiment table, but it is really hard. Since we implemented experiment marking in telemetry, Telemetry has added another new feature: Events. We should put enrollment and unenrollment, along with details about why the user enrolled/unenrolled in this new API.

**Comment by sunahsuh** [comment]
Here's the API to register and record events: https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/events.html#the-api

**Comment by sunahsuh** [comment]
And some lockbox examples:
https://github.com/mozilla-lockbox/lockbox-extension/blob/master/src/webextension/telemetry.js
https://github.com/mozilla-lockbox/lockbox-extension/blob/3482d7e1d245151ae94fa0cf400902c60d8f3f69/src/bootstrap.js#L25

**Comment by mythmon** [comment]
@sunahsuh @jaredkerim @MattGrimes Can you take a look at the below and add any comments you have?

Here is an attempt at spec'ing this out. As a note, the [Telemetry Events API](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/events.html#the-api) has this signature for recording events:

```js
Services.telemetry.recordEvent(category, method, object, value, extra);
```

`category`, `method`, and `object` must be known statically. That is, they can't be the name or id of the recipe, sorry.

---

When enrolling a user in a preference study, this event will be recorded:

```js
Services.telemetry.recordEvent(
  "shield",
  "enroll",
  "preference-study",
  studySlug,
  { isHighPop, branch }
);
```

When unenrolling:

```js
Services.telemetry.recordEvent(
  "shield",
  "unenroll",
  "preference-study",
  studySlug,
  { reason, resetValue }
);
```

`reason` will be one of "filter-mismatch", "user-preference-change-sideload", "user-preference-change". `resetValue` will be the string "true" if Shield reset the value of the preference, and the string "false" otherwise. (Telemetry Event values must be strings here, sorry). Shield does not reset the value when the reason for unenrollment was the preference changing. It does if the reason is the filter no longer matching.

---

For add-on studies (aka "opt-out studies"), the events are similar. The only different is that "object" argument will be "addon-study" instead of "preference-study", and that the keys in the "extra" field will be different.

For enrollment, the extra keys will be `addonId` and `addonVersion`.

For unenrollment, the extra keys will be only include `reason`. The values of `reason` are similar: "filter-mismatch", "uninstalled-sideload", and "uninstalled".

---

Note: "Sideload" here means changes that occur when Shield "isn't watching". This generally refers to changes made to the profile while Firefox is shut down, or sometimes changes that happen due to version upgrades.

**Comment by jaredkerim** [comment]
@mythmon Is it possible to also send the variant the user was enrolled in as one of the extra fields?

**Comment by mythmon** [comment]
@jaredkerim If I understand your question correctly, that is the `branch` field for preference experiments. If you mean the variant chosen by a study add-on, then no we can't include that. That selection process isn't something Shield has any knowledge of. It would be up to individual studies to send events related to that.

**Comment by jaredkerim** [comment]
@mythmon Yes branch is exactly what I meant, it's already there, that's perfect thank you!

**Comment by gregglind** [comment]
Feedback:
1.  Overall, looks good.
2.  Is there room for VERSION (of this API) in there somewhere?  I am nervous about how to make this easy to consume at the back end.  
3.  Want help with jsonSchema for this, so that all ENUM's here are listed?  

I am also very interested in moving all shield stuff to look like this.

**Comment by mythmon** [comment]
I don't think adding a version to this right now is a good idea. There isn't an obvious place to put it, and it isn't clear what would change over time. I'd like to leave version out for now, and then introduce a version field as a "v2" when it becomes needed and we know *what* is needed. This is a strategy I've seen work well in a lot of places.

I'm not interested in verifying this with json schema at this point. Shield isn't building a generic wrapper around Telemetry events, and we already conform to the schema required by Telemetry. I think that adding some sort of schema validation here will be more error prone than testing the code directly.

> I am also very interested in moving all shield stuff to look like this.

I'm not quite sure what you mean by this, but I'd like to take this a bit slowly. I don't want to get into another case where we send too much data too fast (experiment attribution) or record data that we don't end up using (uptake telemetry). Lets focus on landing enroll and unenroll events, and see what happens from that.

**Comment by sunahsuh** [comment]
This looks great -- fits in really well with the intent of each field. Just one note that the "extra" field should be key/value even when recording.

> I'm not interested in verifying this with json schema at this point.

FWIW, if @gregglind is talking about writing a schema for server-side use, that should be unnecessary -- these events should start showing up in main_summary and the exploded Events table as soon as normandy starts sending them without any additional server-side work.

**Comment by mythmon** [comment]
Thanks for looking over this everyone. I'm going to start implementing this today.

> Just one note that the "extra" field should be key/value even when recording.

Thanks for clarifying this. This is my intent. I am using a ES6-shorthand for object construction, in which `{ x, y }` is equivalent to `{ x: x, y: y }`.

**Comment by gregglind** [comment]
Re schemes:  for analysis, what are all the types and enum choices for each
field. It’s a kindness to the analyst.

**Comment by mythmon** [comment]
I've put [a PR for this](https://github.com/mozilla/normandy/pull/1154). The schema I outlined above proved to be not quite realistic, so the actual schema implemented is slightly different. Specifically, `isHighPop` was replaced by `experimentType`, which takes the same value as experimentType in the main ping.

[This preview of the new documentation](https://github.com/mythmon/normandy/blob/32b3682772ca6e8f579e975d185cfdef91f45388/docs/user/data_collection.rst#enrollment) should be useful. @sunahsuh, @jaredkerim, @gregglind can you take a look at it? Does this meet your needs? I specifically tried to enumerate all the values of the enum-like fields for Gregg.

**Comment by gregglind** [comment]
Thank you for enumerating the cases.  I will dig into this!

**Comment by jaredkerim** [comment]
@mythmon Yep that schema looks like exactly what we need, this is super helpful, thank you!

**Comment by mythmon** [comment]
Fixed by #1139.
